docs/api/cypress-api/ensure.mdx

```diff
@@ -0,0 +1,81 @@
+---
+title: Cypress.ensure
+---
+
+`Cypress.ensure.method()` is a collection of helper methods for verifying
+conditions while executing Cypress commands. They are mostly useful when writing
+[custom queries](/api/cypress-api/custom-queries) or
+[commands](/api/cypress-api/custom-commands).
+
+Most functions on `Cypress.ensure` accept `subject` argument, and assert
+something about it, throwing an error if it doesn't match. They have no return
+value; their only purpose is to ensure the state of the application and throw an
+error if it doesn't match.
+
+## Syntax
+
+```javascript
+// Command chain
+Cypress.ensure.isChildCommand(command, args, cy)​
+
+// Type of argument
+Cypress.ensure.isType(subject, type, name, cy)​
+Cypress.ensure.isElement(subject, name, cy)​
+Cypress.ensure.isWindow(subject, name, cy)
+Cypress.ensure.isDocument(subject, name, cy)​
+
+// State of DOM element
+Cypress.ensure.isAttached(subject, name, cy)​
+Cypress.ensure.isNotDisabled(subject, name)​
+Cypress.ensure.isNotHiddenByAncestors(subject, name)​
+Cypress.ensure.isNotReadonly(subject, name)​
+Cypress.ensure.isScrollable($el, name)​
+Cypress.ensure.isStrictlyVisible(subject, name)​
+Cypress.ensure.isVisible(subject, name)​
+```
+
+:::caution
+
+Many of these functions accept an optional `onFail` argument. This is a legacy
+feature used to customize the thrown error, and may be removed in a future
+release; we recommend against relying on it. If you need more control over the
+error thrown, write your own `ensure` function instead.
+
+:::
+
+### Usage
+
+**<Icon name="check-circle" color="green" /> Correct Usage**
+
+```javascript
+Cypress.Commands.addQuery('getChildById', function (id) {
+  return (subject) => {
+    // Verify that the subject is an element, document, or window object
+    Cypress.ensure.isType(subject, ['element', 'document', 'window'], 'getChildById', cy)
+
+    return $$(`#${id}`, subject)
+  }
+})
+
+Cypress.Commands.addQuery('verifyElementActionable', function (id) {
+  // Throws an early error if this query is chained directly off `cy`, eg.
+  // `cy.verifyElementActionable('foo')` would throw an error
+  Cypress.ensure.isChildCommand(command, args, cy)​
```

Comment:
[Here](https://github.com/cypress-io/cypress/pull/25674#discussion_r1093771328) you said query's don't have a concept of parent/child commands. Is this a realistic example? Also where is ` command`, `args`, and `cy` coming form?

Reply:
Fixed the example to show where they come from.

Yes it's a realistic example; it's exactly how we use this function in our own code. I'll update that other file to be clear that queries don't *by default* have a concept of being child / parent - you can make them have one by using `ensure.isChildCommand()`.

Reply:
>  It's just not baked into the query API the way it's baked into the command API.

Gottchya - when you said it didn't have the concept of it, I had understood it as "parent" / "child" truely was no longer a concept vs a difference in how the concept was implemented.

Reply:
tl;dr - it *shouldn't* still be a concept, but it leaked into queries through a new implementation. :P